rclpy/rclpy/parameter_event_handler.py

```diff
@@ -0,0 +1,367 @@
+
+            with self.mutex:
+                self.parameter_callbacks[(parameter_name, node_name)].insert(0, handle)
+
+            return handle
+
+        def remove_parameter_callback(
+            self,
+            handle: ParameterCallbackHandle,
+        ) -> None:
+            """
+            Remove the parameter callback related to provided handle.
+
+            :param handle: A handle of the callback that is to be removed
+            """
+            with self.mutex:
+                handle_key = (handle.parameter_name, handle.node_name)
+                if handle_key in self.parameter_callbacks:
+                    if handle in self.parameter_callbacks[handle_key]:
+                        self.parameter_callbacks[handle_key].remove(handle)
+                    else:
+                        raise RuntimeError("Callback doesn't exist")
+
+                    if len(self.parameter_callbacks[handle_key]) == 0:
+                        self.parameter_callbacks.pop(handle_key)
+                else:
+                    raise RuntimeError("Callback doesn't exist")
+
+        def add_parameter_event_callback(
+            self,
+            callback: Callable[[ParameterEvent], None],
+        ) -> ParameterEventCallbackHandle:
+            """
+            Add new parameter event callback.
+
+            Callbacks are called in FILO manner.
+
+            :param callback: A callable to be referenced on a ParameterEvent
+
+            :return ParameterEventCallbackHandle: The duplicate of a callback handle,
+                stored in a parameter_event dict
+            """
+            handle = ParameterEventCallbackHandle(callback=callback)
+
+            with self.mutex:
+                self.event_callbacks.insert(0, handle)
+
+            return handle
+
+        def remove_parameter_event_callback(
+            self,
+            handle: ParameterEventCallbackHandle,
+        ) -> None:
+            """
+            Remove the parameter event callback related to provided handle.
+
+            :param handle: A handle of the callback that is to be removed
+            """
+            with self.mutex:
+                if handle in self.event_callbacks:
+                    self.event_callbacks.remove(handle)
+                else:
+                    raise RuntimeError("Callback doesn't exist")
+
+    def __init__(
+        self,
+        node: Node,
+        qos_profile: QoSProfile = qos_profile_parameter_events,
+        callback_group: Optional[CallbackGroup] = None,
+        event_callbacks: Optional[SubscriptionEventCallbacks] = None,
+        qos_overriding_options: Optional[QoSOverridingOptions] = None,
+        raw: bool = False,
+    ):
+        """
+        Create ParameterEventHandler.
+
+        Usage example:
+
+        .. code-block:: python
+            import rclpy
+            from rclpy.parameter_event_handler import ParameterEventHandler
+
+            handler = ParameterEventHandler(node, "example")
```

Comment:
This isn't quite right, is it?  That is, I think the second argument should be a `qos_profile`?

Reply:
In this case qos_profile parameter of this class [has the default value](https://github.com/ros2/rclpy/pull/1135/files#diff-8c76dbf5069aaf1524bee18686963328f3039546e2679462d00221815cd029bbR181), so imho, it is not required in this example. But I can add it, if you think it would be better to do so. I based my example on [this one from parameter_client.py](https://github.com/ros2/rclpy/blob/rolling/rclpy/rclpy/parameter_client.py#L60)

Reply:
But my point is that this doesn't work as-is:

```
>>> import rclpy
>>> from rclpy.parameter_event_handler import ParameterEventHandler
>>> rclpy.init()
>>> node = rclpy.create_node('foo')
>>> handler = ParameterEventHandler(node, "example")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "install/rclpy/lib/python3.10/site-packages/rclpy/parameter_event_handler.py", line 235, in __init__
    self.parameter_event_subscription = node.create_subscription(
  File "install/rclpy/lib/python3.10/site-packages/rclpy/node.py", line 1603, in create_subscription
    qos_profile = self._validate_qos_or_depth_parameter(qos_profile)
  File "install/rclpy/lib/python3.10/site-packages/rclpy/node.py", line 1453, in _validate_qos_or_depth_parameter
    raise TypeError(
TypeError: Expected QoSProfile or int, but received <class 'str'>
```

That is, I think this should just be:

```suggestion
            handler = ParameterEventHandler(node)
```

Reply:
Oh, sorry. You are absolutely right.